### ISSUE TITLE
preLaunchTask background begin/ends pattern regex not matching

### ISSUE THREAD

**Issue by zfields** [microsoft/vscode#143941]
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version:
- OS Version:

Version: 1.60.1
Commit: 83bd43bc519d15e50c4272c6cf5c1479df196a4d
Date: 2021-09-10T17:06:57.931Z
Electron: 13.1.8
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Linux x64 5.15.23-76051523-generic

Steps to Reproduce:

1. Create a `launch.json` with a `preLaunchTask` property.

`launch.json`
```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
			"name": "ESP-Prog Debug",
			"type": "gdb",
			"request": "launch",
			"target": "./build/esp_prog_vscode_debug.elf",
			"cwd": "${workspaceRoot}/Radar",
			"env": {
				"PATH": "${config:esp_toolchain_paths}:${env:PATH}"
			},
			"gdbpath": "xtensa-esp32-elf-gdb",
			"preLaunchTask": "Launch OpenOCD GDB Server",
			"autorun": [
				"target extended-remote :3333",
				"mon reset halt",
				"flushregs",
				"thb app_main",
				"c",
				"monitor [target current] configure -event gdb-detach { shutdown }",
			],
		}
    ]
}
```

`tasks.json`
```json
...
        {
            "label": "Launch OpenOCD GDB Server",
            "type": "shell",
            "isBackground": true,
            "options": {
                "cwd": "${workspaceRoot}/Radar",
                "env": {
                    "PATH": "${config:esp_toolchain_paths}:${env:PATH}"
                }
            },
            "command": "openocd -f interface/ftdi/esp32_devkitj_v1.cfg -f target/esp32.cfg",
            "problemMatcher": {
                "pattern": {
                    "regexp": "________",
                    "file": 1,
                    "location": 2,
                    "message": 3
                },
                "background": {
                    "activeOnStart": true,
                    "beginsPattern": ".",
                    "endsPattern": "."
                } 
            }
        },
...
```

The terminal output of the preLaunchTask looks like this...

```none
Open On-Chip Debugger  v0.11.0-esp32-20211220 (2021-12-20-15:42)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
adapter speed: 20000 kHz

Info : Listening on port 6666 for tcl connections
Info : Listening on port 4444 for telnet connections
Info : ftdi: if you experience problems at higher adapter clocks, try the command "ftdi_tdo_sample_edge falling"
Info : clock speed 20000 kHz
Info : JTAG tap: esp32.cpu0 tap/device found: 0x120034e5 (mfg: 0x272 (Tensilica), part: 0x2003, ver: 0x1)
Info : JTAG tap: esp32.cpu1 tap/device found: 0x120034e5 (mfg: 0x272 (Tensilica), part: 0x2003, ver: 0x1)
Info : starting gdb server for esp32.cpu0 on 3333
Info : Listening on port 3333 for gdb connections
```

At first I tried what I thought were reasonable values...

```json
"problemMatcher": {
    "pattern": {
        "regexp": "(Info|Warn) : .*",
        "file": 1,
        "location": 2,
        "message": 3
        },
        "background": {
            "activeOnStart": false,
            "beginsPattern": "Open On-Chip Debugger.*",
            "endsPattern": "Info : Listening on port [0-9]{1,5} for gdb connections"
        }
    }
}
```

and I validated them using `https://regex101.com/`, as suggested in the documentation.

When this didn't work, I discovered myriad issues related to the `preLaunchTask` pop up message, as well as several blog post and Slack Overflow questions. When I couldn't get it to work, I switched to an aggressive (practically useless) problem matcher that was linked multiple times from the dialogues in your issues (shown in full - top), and it doesn't work either.

I don't get it. How does is this supposed to work? Does it have a bug? The documentation online is woefully insufficient, and does not describe in enough depth the parameters and how they interact.

Honestly, I think the only reason you don't get more messages is because you've taught people to ignore the warning and gave them a checkbox to permanently dismiss it.

I read how it is supposed to work, and I think the feature is neat. I have a relatively simple and straight-forward application, and I would like to have it work correctly without generating errors. Can you please help me by breaking down my bad assumptions and/or what I've done wrong in a step-by-step manner? Maybe it will serve as a good example for others who are struggling as much as I am.

**Comment by zfields** [comment]
Yes, I'll gladly update.

Out of curosity, is there a console where I can watch my regex work or fail?

**Comment by zfields** [comment]
I upgraded to 1.64.2, and it immediately failed. I have seen it work (not prompt me to "Debug Anyway") once or twice, but it's inconsistent. I have also seen it just hang.

I'm curious, when you said you tested it, how did you test it exactly? Did you use an Ubuntu machine with with OpenOCD, a connected microcontroller and gdb?

I'm assuming you made a simpler test. If so, can you share your test, and how you confirmed it works and matched the start and end? If you could supply me with that information, then perhaps I could extrapolate how to validate my regex given my actual setup.

**Comment by alexr00** [comment]
@zfields you are correct in assuming I didn't assemble the full setup. Instead, I isolated the task + problem matcher + launch configuration interaction:
- I made a dummy script that echos your real script output. My script is Windows specific, but you get the idea.
- I used whatever launch configuration `type` I already had installed.
- I tested on Windows because tasks are more likely to have issues there due to Windows terminal issues.

I published a repo with my test setup here: https://github.com/alexr00/testPrelaunchBackgroundTask 

I'll try it out on Ubuntu now just in case.

**Comment by alexr00** [comment]
Tested on Ubuntu and update my sample repo to work on linux too. Everything's still working as expected, but I'm interested to see if my sample works for you.

**Comment by zfields** [comment]
I'll be testing it this very soon. I have a busy work week, and this is not on the critical path. Thank you for sharing!

**Comment by zfields** [comment]
What else can I use instead of the launch type "node" for testing?

![image](https://user-images.githubusercontent.com/2801887/156867010-de9172b9-99bf-427e-a753-bb73d4dccceb.png)

I don't have node installed on my machine.

**Comment by weinand** [comment]
After reading you initial comment again I wonder what is exactly the problem?
The title says "... pattern regex not matching"
But how do you know that they are not matching?
What do you expect should happen and what is actually happening?

**Comment by zfields** [comment]
> But how do you know that they are not matching?

Great Question! Unfortunately, I don't know how to test/debug this to know if it's working correctly (feature request), and the only example I was offered requires a dependency I do not have installed. So, I made an assumption based on the error message, behavior, issues on this repository, and multiple blog and stackoverflow posts.

> ... using a "problemMatcher" would not be the correct approach.

Perhaps, but if this is the case, then the following error message is exceedingly poor.

![image](https://user-images.githubusercontent.com/2801887/156939247-bf5590e8-ce61-409d-a92c-68e227ae2b00.png)

I digress...

While testing it this morning, I updated my VSCode from `1.64.2` to `1.65.0`. It looks like whatever the bug was has been fixed in the latest version, and now the `background` object works as you describe and as I was expecting (based on the documentation) and I can no longer reproduce my issue.

```
Version: 1.65.0
Commit: b5205cc8eb4fbaa726835538cd82372cc0222d43
Date: 2022-03-02T11:12:36.248Z
Electron: 13.5.2
Chromium: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Linux x64 5.16.11-76051611-generic
```

I do appreciate everyone's help, thank you for working through this with me. :pray: :heart: :+1:
